orc8r/cloud/go/service/middleware/unary/identity_decorator.go

```diff
@@ -102,82 +101,113 @@ func SetIdentityFromContext(ctx context.Context, _ interface{}, info *grpc.Unary
-			// security hack, either both or neither of the headers should be
-			// set
-			glog.Infof("CCN is present without SCN in metadata: %+v", ctxMetadata)
-			err = status.Error(codes.Unauthenticated, "Inconsistent Request Signature")
+	if snok {
+		newCtx, err = findIdentityFromCSNlist(snlist, ctx, ctxMetadata)
+		if err == nil {
+			return newCtx, newReq, resp, nil
 		}
+	} else if _, ok = ctxMetadata[CLIENT_CERT_CN_KEY]; ok {
+		// CN header is present while SN header is missing - possible
+		// security hack, either both or neither of the headers should be
+		// set
+		err = unexpectedCCN(ctxMetadata)
 	}
 
-	if info != nil {
-		// Check if the call is for a allowlisted method - anything is allowed
-		// do this check past possible identity decoration to still allow to add
-		// valid identity even to allowlisted requests
-		if _, ok := identityDecoratorBypassList[info.FullMethod]; ok {
-			// Bypass method (Bootstrapper & Co.), shortcut...
-			return newCtx, newReq, resp, nil
-		}
+	if isAllowlistedCall(info) {
+		return newCtx, newReq, resp, nil
 	}
 
-	// Only allow local clients if there is no previous errors (snok == false)
 	if err == nil {
 		// We assume that only external calls forwarded by cloud proxy (or unit
 		// tests) will have CSN & CCN headers set. The absence of the headers
 		// along with client IP verification will indicate a local service to
 		// service or Obsidian to service call
-		// For internal calls, no identity verification needed, just make sure
-		// it's a local client
-		err = ensureLocalPeer(ctx)
-		if err != nil {
-			var rpc string
-			if info != nil {
-				rpc = info.FullMethod
-			} else {
-				rpc = "Undefined"
-			}
-			glog.Infof("Empty CTX Metadata from non-local %s client: %v", rpc, err)
-		}
+		err = checkLocalClientCall(ctx, info)
 	}
 	return newCtx, newReq, resp, err
 }
 
+func noMetadata() error {
+	glog.Info(ERROR_MSG_NO_METADATA)
+	return status.Error(codes.Unauthenticated, ERROR_MSG_NO_METADATA)
+}
+
+func unexpectedCCN(ctxMetadata metadata.MD) error {
+	glog.Infof("CCN is present without SCN in metadata: %+v", ctxMetadata)
+	return status.Error(codes.Unauthenticated, "Inconsistent Request Signature")
+}
+
+func multipleCSN(ctxMetadata metadata.MD) error {
+	glog.Infof("Multiple CSNs found in metadata: %+v", ctxMetadata)
+	return status.Error(codes.Unauthenticated, "Multiple CSNs present")
+}
+
+func findIdentityFromCSNlist(snlist []string, ctx context.Context, ctxMetadata metadata.MD) (context.Context, error) {
+	if len(snlist) == 1 {
+		// One CSN is found, find Caller's Identity associated with it
+		newCtx, err := findIdentityFromUniqueCSN(snlist[0], ctx, ctxMetadata)
+		return newCtx, err
+	} else {
+		// there is a certificate serial number (CSN) list in CTX
+		// there can be only one CSN, error out if not
+		return nil, multipleCSN(ctxMetadata)
+	}
+}
+
+func findIdentityFromUniqueCSN(sn string, ctx context.Context, ctxMetadata metadata.MD) (newCtx context.Context, err error) {
```

Comment:
why is `err` passed to that function, if it is not used?

Reply:
It isn't passed to it, it is defined within it. In go you can define variables for the return values in the function declaration like this (the second bracket are the return values, the first bracket are the variables passed to the function), so you can just write e.g. `err = ...` instead of `err := ...` within the function. 
This is also the reason why the variables `gwIdentity` and `certExpTime` are defined the way they are, because `err` is already defined. The main reason to do it like this is probably to have named return values, especially for `newCtx`, because statements like `return nil, nil` are a bit cryptic (or we'd have to define `newCtx` explicitly).
Unfortunately, we have to give names to either all return values or none.

Reply:
Ah sorry, my misstake :). Failed to read the function definition correctly